# 1-36 python基础   

6   
1 没加引号，会误认为是变量。    
2 有引号包裹，被视为字符串,表示文本

In [38]:
len('hello'),'hello'[1]

(5, 'e')

7   
字符串连接+ ；三引号换行

8
等号左边是变量      
great = '吃了吗'    
great_china = great     
great_english = "what'up"   
great = great_english   

9 变量名   
下画线命名法：小写字母由下划线分隔  
不占用关键字

10 math，数学运算库，库里面有很多函数，如sin()。   
默认就有的函数为内置函数，如open()， **print() 默认结尾换行**。   

In [39]:
import math 
a = -1
b = -2
c = 3
x_1 = (-b + math.sqrt(b**2 - 4*a*c))/2*a
x_2 = (-b - math.sqrt(b**2 - 4*a*c))/2*a
# delta = b**2 - 4*a*c
# x_2 = (-b - math.sqrt(delta))/2*a
x_1,x_2

(-3.0, 1.0)

12数据类型，知道所属类型才能使用相应的函数      
空值类型只有None,表示没有值。   
如果暂时不知道赋什么值，可以用None     
my_wife = None

14 input() **返回的是字符串**

In [40]:
user_age = int(input('请输入年龄'))
user_age_10_years = user_age + 10
print('你10年后' + str(user_age_10_years) + '岁了！')

ValueError: invalid literal for int() with base 10: ''

17 逻辑运算符顺序 not > and >or

18 列表[]。shop_list[3]。   可变的；可放置不同类型的数据。       
方法和函数都是实现特定功能，不同在于    
>方法，对象.方法名() ，shop_list.append('显示器'),   shop_list.remove('显示器')      
>>函数, 函数名(对象) , len(shop_list), max(),min(),sorted()


19 字典{}。可变的。     
键，不可变。元组可作为键，因为元组tuple，不可变，不能添加，修改等操作。       
**for "键" in 字典**：此操作能告诉你该键是否存在于字典里    
del 字典["键"]， len(字典)

20 for 循环 迭代对象可以是列表、字典、字符串等      
21 while 无限循环 适合次数未知

In [ ]:
total = 0
count = 0
user_input = input("请输入数字(q结束程序)")
while user_input  != "q":
    num = float(user_input)
    total = total + num
    count = count + 1
    user_input = input("请输入数字(q结束程序)")
if count == 0:
    result = 0
else:
    result = total/count
print(f'一共输入{count}次')
print(f'输入总和是{total}')
print(f'平均值：{result}')

一共输入3次
输入总和是12.0
平均值：4.0


22 字符串格式化。{:.nf}保留n位小数       
数字也可以用format()方法        


In [ ]:
year = '虎'     
name = '小明'

message ="""    
金{0}敲门，五福临门。   
给{1}拜年了!    
""".format(year,name)    

message ="""    
金{current_year}敲门，五福临门。   
给{receive_name}拜年了!    
""".format(receive_name=name,
            current_year=year)    

message =f"""    
金{year}敲门，五福临门。   
给{name}拜年了!    
"""

23 24 函数
函数里定义的变量都只是局部变量。在函数外面是无法访问的。

In [ ]:
def func():
    a = 3
    print(a)
    return a # 没有return语句，返回值默认为None。
return_value = func()

# 如print() append() 这类函数，返回值默认为None。这类函数一般直接调用，不进行赋值。
# 而len() sum() 此类函数要把调用结果赋值给其他变量。

3


25 模块是包含Python定义和声明的文件。每个.py文件都是一个模块，文件名就是模块名。      
- 一个模块可以包含多个类、函数、变量等。  
- 引入模块后，里面的函数和变量就可以使用。           
- 引入就像是把某个家具从自家仓库拿出来使用,三种引入方法。           
- 模块可以被导入到其他Python脚本中，以便使用其定义的功能。  

26 面向对象编程（封装。继承。多态：可以调用同一名称的方法，但由于所属具体子类的不同，执行不同的内容）              
>面向对象编程以对象(人、洗衣机)为核心，先考虑各个对象有什么性质(属性，类里面的变量)，能做什么事情(方法，类里面的函数)。   
>>把对象定义为类，可以作为函数的参数传入。      
>>>类定义了对象的属性和方法（模板）；而实例化后的不同对象拥有的具体属性和方法可以不一样。      
>>>>用对象把相关属性和方法绑定在一起，更有利于程序逻辑清晰

- 面向过程编程类似函数，把要完成事情拆分成一个个步骤，按步骤依次完成。     

27 对象属性

In [ ]:
class Cutecat:
    def __init__(self,cat_name,age,color):
        # 以下是构造函数的代码，__init__()定义对象有哪些属性.
        # self表示对象自身，可以把属性的值绑定在实例对象上。
        self.name = cat_name # 左边代表对象的属性，被绑定为传入参数的值.
        self.age = age
        self.color = color

cat1 = Cutecat('Jojo',6,'blue')
print(f'小猫{cat1.name}的年龄是{cat1.age}岁,花色是{cat1.color}')

小猫Jojo的年龄是6岁,花色是blue


28 对象方法

In [ ]:
# 实例化对象时，init会自动运行.
class Student:
    def __init__(self,name,Student_id): # self把属性的值绑定在实例对象上
        self.name = name #左边是绑定到对象上的属性，右边是普通变量，由传入参数决定
        self.student_id = Student_id
        self.grades = {'语文':0,'数学':0,'英语':0} #如果每个对象拥有同样的初始值，就不需要从参数去获取

    def set_grade(self,course,grade): # 方法可以获取or修改，和对象绑定的属性
        if course in self.grades:
            self.grades[course] = grade

    def print_grades(self):
        print(f'学生:{self.name} (学号:{self.student_id}) 的成绩为:')
        for course in self.grades:
            print(f'{course}:{self.grades[course]}分')


chen = Student('小陈','10086')
chen.set_grade('语文',92)
chen.set_grade('数学',94)
chen.print_grades()
# zeng = Student('小曾','10087')
# chen.name,zeng.grades
# zeng.set_grade('数学',95)
# zeng.grades

学生:小陈 (学号:10086) 的成绩为:
语文:92分
数学:94分
英语:0分


29 类继承：子类会继承父类的属性和方法       
当A是B，classA(B):      
当新能源车是车，class新能源车(车)：

In [ ]:
# 父类
class Mammal:
    def __init__(self,name,sex):
        self.name = name
        self.sex = sex
        self.num_eyes = 2

    def breathe(self):
        print(self.name + '在呼吸')

    def poop(self):
        print(self.name + '在拉屎')


In [ ]:
# 子类
class Human(Mammal):
    def __init__(self,name,sex):
        super().__init__(name,sex)  # 调用父类的的构造函数
        self.has_tail = False       # 不同子类的特有属性

    def read(self):                 # 不同子类的特有方法
        print(self.name + '在阅读')

class Cat(Mammal):
    def __init__(self,name,sex):
        super().__init__(name,sex)
        self.has_tail = True            # 不同子类的特有属性

    def scratch_sofa(self):             # 不同子类的特有方法
        print(self.name + '在抓沙发')

    def poop(self):
        print(self.name + '在猫砂上拉屎') # 若子类和父类有同名方法，则优先调用子类的方法

cat2 = Cat('Jojo', '男')
cat2.scratch_sofa()
cat2.poop()
cat2.has_tail

Jojo在抓沙发
Jojo在猫砂上拉屎


True

练习：人力系统

In [ ]:
class Employee():
    def __init__(self,name,id):
        self.name = name
        self.id = id

    def print_info(self):
        print(f'员工的姓名是:{self.name},员工的工号是:{self.id}')

class FullEmployee(Employee):
    def __init__(self,name,id,monthly_salary):
        super().__init__(name,id)
        self.monthly_salary = monthly_salary

    def calculate_monthly_pay(self):
        return self.monthly_salary

class ParttimeEmployee(Employee):
    def __init__(self,name,id,daily_salary,work_days):
        super().__init__(name,id)
        self.daily_salary = daily_salary
        self.work_days = work_days
    
    def calculate_monthly_pay(self):
        return self.daily_salary * self.work_days

ff = FullEmployee('小明','10086',10000)
pp = ParttimeEmployee('小王','10087',300,22)


In [ ]:
ff.print_info()
pp.print_info()
ff.calculate_monthly_pay(),pp.calculate_monthly_pay()

员工的姓名是:小明,员工的工号是:10086
员工的姓名是:小王,员工的工号是:10087


(10000, 6600)

30 目录。 mac,用/     而win,C:\  D:\   
绝对路径，从根目录出发。C:\home\data\a.py， win用反斜杠进行分割，其它系统用/进行分割。      
相对路径，从自身位置出发，来看其他文件处于什么路径。        

31 读取文件操作 

In [ ]:
f = open('./poem.txt','r',encoding='utf-8')
f.read(10)          # 读取10字节内容，默认是读取全部
line = f.readline() # 每次读取一行内容，含读取换行符。配合while循环
f.readlines()       # 读取全部，返回由每行组成的字符串列表，配合for循环使用

['高处不胜寒。\n', '123\n', '起舞弄清影\n', '何似在人间']

32 写入文件操作   
文件不存在会自动创建；已存在会覆盖。       
r+ 同时读写

In [ ]:
with open('./poem.txt','w',encoding='utf-8') as f:
    f.write('我欲乘风归去,\n又恐琼楼玉宇,\n高处不胜寒。') # 需要手动加换行符
    f.write('\n123\n')

In [ ]:
with open('./poem.txt','a',encoding='utf') as f:
    f.write('起舞弄清影\n')
    f.write('何似在人间')

In [ ]:
with open('./poem.txt','r',encoding='utf-8') as f:
    print (f.read())

我欲乘风归去,
又恐琼楼玉宇,
高处不胜寒。
123
起舞弄清影
何似在人间


33 异常处理     

In [ ]:
try:
    可能会报错的代码
except 错误类型:
    错误发生后想执行的操作
except 错误类型:
    错误发生后想执行的操作
except:                     # 捕捉所有错误类型
    错误发生后想执行的操作
else:
    try语句正确时的操作
finally:
    无论发生错误与否，都要执行的语句。

SyntaxError: invalid character '，' (U+FF0C) (3455420080.py, line 8)

34 测试：验证新代码是否正确，同时不影响旧代码的执行     
assert + bool表达式，为True则继续运行        

In [ ]:
import unittest #内置库
unittest 

<module 'unittest' from 'c:\\Users\\Qjn\\anaconda3\\lib\\unittest\\__init__.py'>

35 测试

36 python高阶和匿名函数     
  
- 高阶函数：**把函数作为参数的函数**。格式化函数formatter也可以作为参数，扩展出不同打印格式。    
`calculate_print = (3,calculate_square,formatter)` **这里不用带()和参数**, 如果带()就是在调用函数，传入的就是函数的执行结果，而不是函数本身。       
- 匿名函数，不需要起名字，像个一次性物品，即用即扔，适用一次性调用的场景。当成没有名字，且占用行数更少的函数。    
>用lambda定义，calculate_print = (3，lambda num1，num2: num1+num2)        
>>冒号前面的变量名，表示传给匿名函数的参数num1，num2；冒号后面直接放上需要返回的结果num1+num2。冒号后面只能是单个表达式..       
>>>匿名函数被定义好后，也可以像普通函数般被直接调用，不同点在于**匿名函数本身也要加括号**。

In [ ]:
(lambda num1,num2: num1+num2)(2,3)   

5

# 37 数据分析环节

38  
shift 选中多个单元格    
单元格交互模式，无需print就能输出       
重启按钮，清空所有定义过的变量      
重启并重新运行所有单元格

- 行内插入公式    $ x^2 + y^2 =1 $        
- 独占一行    $$ x^2 + y^2 =1 $$

39 markdown与latex

我是*斜体字*  
这里~~被删除了~~
1. 列表第一项
2. 第二项
3. 第三项  

[中国气象局](https://weather.cma.cn/)  
![插入图片](https://cn.bing.com/images/search?view=detailV2&ccid=B8yw5AKE&id=2C16862AD728F6D08797DB0F68E471ED0D47435A&thid=OIP.B8yw5AKEVbWuyiSekZctbQHaEK&mediaurl=http://pic.weather.com.cn/images/cn/photo/2022/05/10/20220510161042C24F50226927BFB58907E655DF60A180.jpg&q=%E5%A4%A9%E6%B0%94&ck=08DA7197F01821B1F60C12D56E241784&idpp=rc&idpview=singleimage&form=rc2idp&ajaxhist=0&ajaxserp=0)

文字里插入代码  
```python
print('hello')
```

Latex  
$$x \times y$$
$$x \div y$$
$$ H_2O $$
$$ H_{input} $$
$$ \sum(x^2 + y^2) $$
$$ \sqrt[3]{a^2m^2} $$
$$ \sqrt[3]a^2m^2 $$
$$ \frac{x+y}{x-y} $$

# 40-43  numpy   
numpy数组和列表相似之处在于：
1. 切片
2. 索引
3. 迭代  
不同在于：numpy要求数据类型一致；列表则可以是多种数据类型混合。

In [ ]:
import numpy as np
arr1 = np.array([7,8,9])
print(arr1)
arr2 = np.array([[1,2,3,],[4,5,6]])
arr2

[7 8 9]


array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
print(arr1.ndim) # 维度
print(arr2.ndim)
print(arr1.shape)
print(arr2.size) # 元素个数
print(arr2.dtype)

1
2
(3,)
6
int32


In [ ]:
# 其他创造数组的方法
print(np.zeros(3))
print(np.ones(6))
print(np.arange(5,10,2)) # 区分与python的range

[0. 0. 0.]
[1. 1. 1. 1. 1. 1.]
[5 7 9]


In [ ]:
arr3 = np.ones(2)
arr4 = np.arange(5,15,2)
np.concatenate((arr3,arr4))
np.concatenate([arr3,arr4])


array([ 1.,  1.,  5.,  7.,  9., 11., 13.])

- python  
内置函数sorted()

In [ ]:
list = [5,17,3,26,31]
sorted_list1 =sorted(list) # 内置函数,传入列表后，就会返回一个新的排序。
print(sorted_list1)        
list                       # sorted调用后只返回排序好的新列表，不改变原始列表。

[3, 5, 17, 26, 31]


[5, 17, 3, 26, 31]

数组自己的方法,sort()

In [ ]:
list = [5,17,3,26,31]
list.sort()                # 数组自己的方法,sort调用后什么都不返回，但原始列表会被排序好。
print(list) 

[3, 5, 17, 26, 31]


- numpy  
numpy的函数np.sort()

In [ ]:
arr5 = np.array([5,17,3,26,31])
print(np.sort(arr5))        # np.sort()调用后只返回排序好的新列表，不改变原始列表。
arr5

[ 3  5 17 26 31]


array([ 5, 17,  3, 26, 31])

数组自己的方法,sort()

In [ ]:
arr5 = np.array([5,17,3,26,31])
arr5.sort()                 # 数组自己的方法,sort()调用后什么都不返回，但原始列表会被排序好。
print(arr5)

[ 3  5 17 26 31]


In [ ]:
print(100*arr5) # 看成标量乘向量,广播机制
print(arr4+arr5) 
print(arr4-arr5)
print(arr4*arr5)
print(arr4/arr5)

[ 300  500 1700 2600 3100]
[ 8 12 26 37 44]
[  2   2  -8 -15 -18]
[ 15  35 153 286 403]
[1.66666667 1.4        0.52941176 0.42307692 0.41935484]
31


In [ ]:
# 聚合操作,由一组值得到一个值
print(arr5.max()) # min() sum() mean()

31


In [ ]:
# 筛选
arr5[(arr5>15) & (arr5<30)]

array([17, 26])

# 44 pandas    
44-45 
- 构建在numpy之上,继承了其高性能的数组计算过程;还提供复杂的数据处理功能.  
- .values返回的类型是numpy,所以针对numpy数组的操作都可以用  

**Series,切片和索引用loc和iloc,以免混淆**
> loc 用标签索引,取值或切片
> >iloc 用原始位置索引,取值或切片

In [47]:
import pandas as pd
s1 = pd.Series([5,17,3,26,31])
print(s1)
print(s1.values)
print(s1.index)

0     5
1    17
2     3
3    26
4    31
dtype: int64
[ 5 17  3 26 31]
RangeIndex(start=0, stop=5, step=1)


In [70]:
s2 = pd.Series([5,17,3,26,31],index=['a','b','c','d','e'])
s2 = pd.Series({'a':5,'b':17,'c':3,'d':26,'e':31})         # 传入字典构造法,键为标签
print(s2)

print(s2['c'] == s2[2]) # 多了一种索引方法.
s2['c':'e'] # 标签索引的切片,结束值是包含的

a     5
b    17
c     3
d    26
e    31
dtype: int64
True


c     3
d    26
e    31
dtype: int64

用索引获取任意元素

In [69]:
s2[[0,4,3]] 
s2[['a','e','d']] 

a     5
e    31
d    26
dtype: int64

整数也可以作为标签  

In [74]:
s3 = pd.Series([5,17,3,26,31],index=[1,3,5,7,9])
s3

1     5
3    17
5     3
7    26
9    31
dtype: int64

In [78]:
s3[(s3>3) & (s3<30)]

1     5
3    17
7    26
dtype: int64

46-48 
- Series加减乘除计算,会根据索引自动对齐去计算,未匹配的索引结果是NaN  
- Series计算,建议使用方法 sub() mul() div()  `s1.add(s2,fill_value=0)`  0代替未找到对应索引的值
- 和numpy一样,也有max(),min(),sum() mean();还有describe()
- 也有广播机制,s1*3 
- apply()方法以函数为参数,相当于是高阶函数, s1.apply(func)调用时,把s1里的每个元素分别作为func()的参数.返回一个新的Series且不改变原始Series
- 函数逻辑比较简单时,可以使用匿名函数作为func,比如s1.apply(lambda x: x+1)

49 dataframe就像是由Series组成的字典,每个Series对应一个键名也就是列名
- dataframe的列是Series类型

- 方法一
- dataframe默认索引和series一样,都是从零开始依次递增的整数来表示位置，  
- 但如果传入的series本身有标签索引的话,dataframe的索引也会变成相应的标签索引

In [76]:
s_id = pd.Series(["01", "02", "03", "04", "05"], 
                 index=["小明", "小红", "小杰", "小丽", "小华"])
s_class = pd.Series(["二班", "一班", "二班", "三班", "一班"], 
                    index=["小明", "小红", "小杰", "小丽", "小华"])
s_grade = pd.Series([92, 67, 70, 88, 76], 
                    index=["小明", "小红", "小杰", "小丽", "小华"])

df3 = pd.DataFrame({"学号": s_id, "班级": s_class, "成绩": s_grade})
df3

学号  班级  成绩
小明  01  二班  92
小红  02  一班  67
小杰  03  二班  70
小丽  04  三班  88
小华  05  一班  76

- 方法二
- 如果想一次性创建出既有标签索引也有列名的dataframe, 可以传入一个嵌套字典，也就是字典里面包含字典.  
- 最外层的键仍然对应各个列名, 而值则对应每列的series, 里层字典的键对应series的标签索引

In [77]:
df4 = pd.DataFrame({"学号": {"小明": "01", "小红": "02", "小杰": "03", 
                           "小丽": "04", "小华": "05"}, 
                    "班级": {"小明": "二班", "小红": "一班", "小杰": "二班", 
                           "小丽": "三班", "小华": "一班"},
                    "成绩": {"小明": 92, "小红": 67, "小杰": 70, 
                           "小丽": 88, "小华": 76}})
df4

学号  班级  成绩
小明  01  二班  92
小红  02  一班  67
小杰  03  二班  70
小丽  04  三班  88
小华  05  一班  76

- 提取对应列   
df['列名'],   df.列名 如果**列名里面有空格或特殊符号**,不能通过属性名来获取,只能通过方括号

- 提取对应行  
loc[]与iloc[]

- 提取对应表格数据
> df.loc['行名','列名'],df.loc['小红':'小杰','班级':'成绩']
>> df.iloc[1:3,1:3]

51

- 更新列or创建列

In [ ]:
df1["成绩"] = pd.Series([90, 91, 83, 95, 94, 85], 
                      index=["001", "002", "003", 
                             "004", "005", "006"])   # Series的索引要一致
df1

In [ ]:
df1["成绩"] = pd.Series([90, 91, 83, 95, 94, 85])    # 用列表也可以,自动按顺序对齐
df1

- 更新行,loc和iloc
- 添加行,只能loc

- 删除行or列,使用 **.drop()方法**,注意axis=0和axis=1的问题  
- `df1.drop(["身高", "性别"], axis=1)`

- DataFrame加减乘除计算,会根据索引自动对齐去计算,未匹配的索引结果是NaN  
- DataFrame计算,建议使用方法 sub() mul() div()  ,  `s1.add(s2,fill_value=0)`  0代替未找到对应索引的值

- DataFrame也可以和Series进行操作,**列名和Series的索引**对齐

In [85]:
df9 = pd.DataFrame({"a": {"001": 0, "003": 4, "005": 8, "007": 12},
                    "b": {"001": 1, "003": 5, "005": 9, "007": 13},
                    "c": {"001": 2, "003": 6, "005": 10, "007": 14},
                    "d": {"001": 3, "003": 7, "005": 11, "007": 15}})
df9

a   b   c   d
001   0   1   2   3
003   4   5   6   7
005   8   9  10  11
007  12  13  14  15

In [86]:
s9 = pd.Series([0.1, 0.2, 0.3, 0.4, 0.5], index=["a", "b", "c", "d", "e"])
s9

a    0.1
b    0.2
c    0.3
d    0.4
e    0.5
dtype: float64

In [87]:
s9*df9

a    b    c    d   e
001  0.0  0.2  0.6  1.2 NaN
003  0.4  1.0  1.8  2.8 NaN
005  0.8  1.8  3.0  4.4 NaN
007  1.2  2.6  4.2  6.0 NaN

53  
- 沿着索引纵轴操作,axis=0,默认    
- 沿着列名横轴操作,axis=1

- DataFrame也有apply()方法 df.apply(func)调用时,运用在每行或每列上
- applymap()方法,运用在每个元素上

54 练习

## 1. 创建一个叫`student_average`的DataFrame，索引为学号，包含姓名列（列名为“姓名”），以及计算出三次考试平均值后的结果列（列名为“平均分”）。

In [90]:
students = pd.DataFrame({ "001": {"姓名": "小陈", "考试1": 85, "考试2": 95, "考试3": 92}, 
                          "002": {"姓名": "小李", "考试1": 91, "考试2": 92, "考试3": 94},  
                          "003": {"姓名": "小王", "考试1": 86, "考试2": 81, "考试3": 89}, 
                          "004": {"姓名": "小张", "考试1": 79, "考试2": 89, "考试3": 95},  
                          "005": {"姓名": "小赵", "考试1": 96, "考试2": 91, "考试3": 91}, 
                          "006": {"姓名": "小周", "考试1": 81, "考试2": 89, "考试3": 92} }).T
students

姓名 考试1 考试2 考试3
001  小陈  85  95  92
002  小李  91  92  94
003  小王  86  81  89
004  小张  79  89  95
005  小赵  96  91  91
006  小周  81  89  92

In [114]:
average1 = students.loc[:, "考试1":"考试3"].mean(axis=1)
name1 = students["姓名"]

姓名        平均分
001  小陈  90.666667
002  小李  92.333333
003  小王  85.333333
004  小张  87.666667
005  小赵  92.666667
006  小周  87.333333

方法1：在合并之前为Series指定列名

In [116]:
average2 = pd.Series(average1,name="平均分")
name2 = pd.Series(name1,name='姓名')

student_average = pd.concat([name2,average2],axis=1)
student_average

姓名        平均分
001  小陈  90.666667
002  小李  92.333333
003  小王  85.333333
004  小张  87.666667
005  小赵  92.666667
006  小周  87.333333

方法2：在合并之后为DataFrame指定列名

In [ ]:
student_average = pd.concat([name1,average1],axis=1)
student_average.columns = ['姓名', '平均分']
student_average

姓名        平均分
001  小陈  90.666667
002  小李  92.333333
003  小王  85.333333
004  小张  87.666667
005  小赵  92.666667
006  小周  87.333333

In [118]:
student_average = pd.DataFrame({"姓名":name1,"平均分":average1}) # 创建方法一
student_average

姓名        平均分
001  小陈  90.666667
002  小李  92.333333
003  小王  85.333333
004  小张  87.666667
005  小赵  92.666667
006  小周  87.333333

## 输出每次考试第二高的分数组成的Series，可以尝试结合匿名函数。

In [128]:
students.loc[:, "考试1":"考试3"].apply(lambda x:np.sort(x)[-2],axis=0) 

考试1    91
考试2    92
考试3    94
dtype: int64

## 根据以下规则，输出每个学号对应的四次考试的等级：
- A+：平均分 >= 95
- A：90 <= 平均分 < 95
- B+：85 <= 平均分 < 90
- B：80 <= 平均分 < 85
- C+：75 <= 平均分 < 80
- C：平均分 < 75

In [135]:
def grade_rank(x):
    if x >= 95:
        return 'A+'
    if x >= 90:
        return "A"
    if x >= 85:
        return "B+"
    if x >= 80:
        return "B"
    if x >= 75:
        return "C+"
    else:
        return "C"
    
students_rank = students.iloc[:,1:].applymap(grade_rank)
students_rank 

考试1 考试2 考试3
001  B+  A+   A
002   A   A   A
003  B+   B  B+
004  C+  B+  A+
005  A+   A   A
006   B  B+   A

In [137]:
students.iloc[:,1:].describe()

考试1  考试2  考试3
count     6    6    6
unique    6    5    5
top      85   89   92
freq      1    2    2

## 展示`students`每个考试分数数据列的统计信息，包括个数、平均值、标准差、四分位数等。

In [138]:
students.describe()

姓名  考试1  考试2  考试3
count    6    6    6    6
unique   6    6    5    5
top     小陈   85   89   92
freq     1    1    2    2

In [140]:
students['考试1'] 
# 字符串数字都是object类型(很宽泛)的子类,这里的数字列是object,就不会展示统计信息
# 既有数字列,又有非数字列的dataframe,describe()才会自动忽略非数字列

001    85
002    91
003    86
004    79
005    96
006    81
Name: 考试1, dtype: object

In [144]:
students2 = students
students2['考试1'] = students2['考试1'].astype("int")
students2['考试2'] = students2['考试2'].astype("int")
students2['考试3'] = students2['考试3'].astype("int")
students2.describe()

考试1        考试2        考试3
count   6.000000   6.000000   6.000000
mean   86.333333  89.500000  92.166667
std     6.314006   4.722288   2.136976
min    79.000000  81.000000  89.000000
25%    82.000000  89.000000  91.250000
50%    85.500000  90.000000  92.000000
75%    89.750000  91.750000  93.500000
max    96.000000  95.000000  95.000000

# Part 3 数据分析流程
- 获取数据，读取数据，评估数据，清洗数据，整理数据，分析数据，可视化数据。  
  
55 获取数据
- 飞桨,天池,和鲸
- 爬虫:网页源码,需要解析
- API:直接获得数据  


56 常见数据格式  json，纯文本
- json有两种数据结构,对象和数组,可以分别被转换成py字典和列表。一个对象是一个示例，对应dataframe的一行
>对象以大括号{}开头和结尾，然后里面都是键值对，每个键值对之间用逗号进行分隔。不要求所有对象的键都相同
>>数组以中括号[]开头和结尾，然后里面是一个个值，每个值之间用逗号进行分隔.

- json有一些额外规则:
>比如python,字典可以用**整数等不可变数据类型**作为键。但json数据里面只能是**字符串作为键，键不能重复,而且字符串必须被双引号包围，单引号不行。**
>>另外，这个规则不仅针对外层的键，如果json对象或数组里面又嵌套了对象，那规则也同样适用于里面的键。  

json值的类型需要属于以下这6种。
- 第一是字符串，注意用双引号包围。
- 第二是整数或浮点数都行。
- 第三是布尔值，也就是true或者false.因为它本质是JavaScript的对象，而JavaScript的布尔值是以小写开头。
- 第四是数组，json对象的值可以是数组，json数组的值也可以是数组。支持嵌套就可以支持复杂的数据格式。
- 第五是对象，对象的值可以是对象;数组里面的值也可以是对象。
- 第六是空值null，你可以把它看成是py的None，总之都是用来表示值不存在.

57 读取数据

In [7]:
import pandas as pd
cellphone_df = pd.read_json(r"D:\data analysis\LinLiLi\03数据获取与读取篇\cell_phones_survey.json") # 数组
cellphone_df

questionType        asin    answerTime    unixTime  \
0       yes/no  1466736038   Mar 8, 2014  1394265600   
1       yes/no  1466736038  Jan 29, 2015  1422518400   
2       yes/no  1466736038  Nov 30, 2014  1417334400   
3   open-ended  1466736038   Nov 3, 2014  1415001600   
4       yes/no  1466736038   Oct 2, 2014  1412233200   
5   open-ended  1466736038  Sep 11, 2014  1410418800   
6       yes/no  1621911888  Dec 13, 2013  1386921600   

                                            question answerType  \
0                         Is there a SIM card in it?          Y   
1  Is this phone new, with 1 year manufacture war...          ?   
2  can in it be used abroad with a different carr...          Y   
3                      What is the warranty on this?        NaN   
4  Does this phone use the regular Sim card (the ...          ?   
5  how much time you need to send me this product...        NaN   
6                                    Is it unlocked?          Y   

                                              answer  
0  Yes. The Galaxy SIII accommodates a micro SIM ...  
1  It is new but I was not able to get it activat...  
2                                                Yes  
3                                        No warranty  
4                                  it takes mini sim  
5  If you choose expedited shipping you will have...  
6                                                yes

In [8]:
git_df = pd.read_json(r"03数据获取与读取篇/github.json") # 对象
git_df
# 此处language的值是长度为3的数组，当键值对里的值为数组时，pandas会把数组元素视为不同数据示例，分别拆分成单独一行，因此language属性下面有三行。

owner           name  stars  forks  watchers  isFork  \
0  pelmers  text-rewriter     11      4         3   False   
1  pelmers  text-rewriter     11      4         3   False   
2  pelmers  text-rewriter     11      4         3   False   

                               languages  \
0  {'name': 'JavaScript', 'size': 21769}   
1         {'name': 'HTML', 'size': 2096}   
2          {'name': 'CSS', 'size': 2081}   

                                description             createdAt  \
0  Webextension to rewrite phrases in pages  2015-03-14T22:35:11Z   
1  Webextension to rewrite phrases in pages  2015-03-14T22:35:11Z   
2  Webextension to rewrite phrases in pages  2015-03-14T22:35:11Z   

               pushedAt  license  
0  2022-02-11T14:26:00Z      NaN  
1  2022-02-11T14:26:00Z      NaN  
2  2022-02-11T14:26:00Z      NaN

58 常见数据格式 CSV，逗号分隔值，纯文本  
59 读取数据


 


In [17]:
# 默认第一行为列名，可header = none
# 但不会默认第一列为索引，可index_col='列名'。默认是位置索引
fifa_df = pd.read_csv(r'D:\data analysis\LinLiLi\03数据获取与读取篇\fifa_players.csv')
# pd.set_option('display.max_columns',120)
# pd.set_option('display.max_colwidth',120) 
fifa_df.tail(5)

C:\Users\Qjn\AppData\Local\Temp\ipykernel_22292\198900098.py:3: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  fifa_df = pd.read_csv(r'D:\data analysis\LinLiLi\03数据获取与读取篇\fifa_players.csv')


player_id                               player_url  fifa_version  \
161578     269011          /player/269011/ziye-zhao/230002            23   
161579     269019          /player/269019/junjie-wu/230002            23   
161580     271093      /player/271093/kailin-barlow/230002            23   
161581     271555      /player/271555/gideon-tetteh/230002            23   
161582     271608  /player/271608/aqeel-al-dhafeeri/230002            23   

        fifa_update fifa_update_date      short_name          long_name  \
161578            2       2022-09-26       Zhao Ziye                赵梓业   
161579            2       2022-09-26       Wu Junjie                吴俊杰   
161580            2       2022-09-26       K. Barlow      Kailin Barlow   
161581            2       2022-09-26       G. Tetteh      Gideon Tetteh   
161582            2       2022-09-26  A. Al Dhafeeri  Aqeel Al Dhafeeri   

       player_positions  overall  potential  value_eur  wage_eur  age  \
161578               RW       46         61   110000.0     700.0   18   
161579               LB       46         58   110000.0     750.0   19   
161580              CDM       46         58   110000.0     500.0   19   
161581           RW, LW       46         70   150000.0     500.0   17   
161582               CB       46         63   110000.0     500.0   17   

               dob  height_cm  weight_kg  league_id       league_name  \
161578  2003-10-01        180         73     2012.0      Super League   
161579  2003-06-02        188         83     2012.0      Super League   
161580  2003-06-28        181         73       65.0  Premier Division   
161581  2005-05-20        175         68       65.0  Premier Division   
161582  2004-07-09        180         70      350.0        Pro League   

        league_level  club_team_id        club_name club_position  \
161578           1.0      112978.0       Hebei CFFC           RES   
161579           1.0      112429.0    Guangzhou R&F           RES   
161580           1.0         563.0     Sligo Rovers           RES   
161581           1.0         306.0  Shamrock Rovers           RES   
161582           1.0      112990.0         Al Batin           RES   

        club_jersey_number club_loaned_from club_joined_date  \
161578                45.0              NaN       2022-05-01   
161579                41.0              NaN       2022-05-01   
161580                34.0              NaN       2021-11-02   
161581                28.0              NaN       2022-06-24   
161582                42.0              NaN       2022-08-26   

        club_contract_valid_until_year  nationality_id     nationality_name  \
161578                          2024.0             155             China PR   
161579                          2027.0             155             China PR   
161580                          2023.0              25  Republic of Ireland   
161581                          2027.0              25  Republic of Ireland   
161582                          2027.0             183         Saudi Arabia   

        nation_team_id nation_position  nation_jersey_number preferred_foot  \
161578             NaN             NaN                   NaN          Right   
161579             NaN             NaN                   NaN           Left   
161580             NaN             NaN                   NaN          Right   
161581             NaN             NaN                   NaN          Right   
161582             NaN             NaN                   NaN          Right   

        weak_foot  skill_moves  international_reputation      work_rate  \
161578          3            2                         1  Medium/Medium   
161579          3            2                         1  Medium/Medium   
161580          2            2                         1    High/Medium   
161581          3            2                         1  Medium/Medium   
161582          3            2                         1  Medium/Medium   

               body_t

In [16]:
fifa_df.info()
fifa_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161583 entries, 0 to 161582
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 135.6+ MB


player_id   fifa_version  fifa_update        overall  \
count  161583.000000  161583.000000     161583.0  161583.000000   
mean   214484.722353      19.125514          2.0      65.699071   
std     34928.608856       2.559318          0.0       7.040855   
min         2.000000      15.000000          2.0      40.000000   
25%    199159.000000      17.000000          2.0      61.000000   
50%    220621.000000      19.000000          2.0      66.000000   
75%    236958.000000      21.000000          2.0      70.000000   
max    271817.000000      23.000000          2.0      94.000000   

           potential     value_eur       wage_eur            age  \
count  161583.000000  1.595300e+05  159822.000000  161583.000000   
mean       70.744008  2.326770e+06   10855.409768      25.123181   
std         6.259121  6.005746e+06   21941.656285       4.670207   
min        40.000000  1.000000e+03     500.000000      16.000000   
25%        66.000000  3.250000e+05    2000.000000      21.000000   
50%        70.000000  7.250000e+05    4000.000000      25.000000   
75%        75.000000  1.800000e+06   10000.000000      28.000000   
max        95.000000  1.940000e+08  575000.000000      54.000000   

           height_cm      weight_kg      league_id   league_level  \
count  161583.000000  161583.000000  159810.000000  158866.000000   
mean      181.240205      75.235031     210.409017       1.380283   
std         6.750148       7.000456     444.685024       0.750646   
min       154.000000      49.000000       1.000000       1.000000   
25%       176.000000      70.000000      19.000000       1.000000   
50%       181.000000      75.000000      56.000000       1.000000   
75%       186.000000      80.000000     308.000000       2.000000   
max       208.000000     110.000000    2149.000000       5.000000   

        club_team_id  club_jersey_number  club_contract_valid_until_year  \
count  159810.000000       159810.000000                   159810.000000   
mean    44914.264783           20.161323                     2020.352418   
std     53413.302607           16.870350                        2.662184   
min         1.000000            1.000000                     2014.000000   
25%       450.000000            8.000000                     2018.000000   
50%      1889.000000           17.000000                     2021.000000   
75%    110832.000000           27.000000                     2022.000000   
max    116361.000000           99.000000                     2032.000000   

       nationality_id  nation_team_id  nation_jersey_number      weak_foot  \
count   161583.000000     9316.000000           9316.000000  161583.000000   
mean        55.570462    30414.309790             12.094354       2.939480   
std         47.891746    48296.823056              6.720156       0.662863   
min          1.000000      974.000000              1.000000       1.000000   
25%         21.000000     1343.000000              6.000000       3.000000   
50%         45.000000     1365.000000             12.000000       3.000000   
75%         56.000000   105035.000000             18.000000       3.000000   
max        219.000000   111527.000000             30.000000       5.000000   

         skill_moves  international_reputation  release_clause_eur  \
count  161583.000000             161583.000000        1.036220e+05   
mean        2.329905                  1.107140        4.797334e+06   
std         0.751792                  0.383605        1.231500e+07   
min         1.000000                  1.000000        9.000000e+03   
25%         2.000000                  1.000000        6.100000e+05   
50%         2.000000                  1.000000        1.300000e+06   
75%         3.000000                  1.000000        3.600000e+06   
max         5.000000                  5.000000        3.735000e+08   

                pace       shooting        passing      dribbling  \
count  143614.000000  143614.000000  143614.000000  143614.000000   
mean       68.0

60  61 评估数据
- 结构：整洁数据(让代码更高效执行)
  1. 每列只包含一个变量
  2. 每行是一个观察值
  3. 每个单元格一个值
- 内容：干净数据，区分于脏数据：
    1. 缺失数据，`.isnull().sum()`。                          **不处理；删除行；平均数等填补**
  1. 重复数据，`duplicated(subset=[])` ,第二次出现为true。   **找到，删除。**
  2. 不一致数据：不同数据值，实际含义相同。`.value_counts()`  **替换成统一**
  3. 无效数据：负身高。`.sort_values()`  **删除or替换**
  4. 错误数据：体重3公斤   **删除or替换**

62 清洗数据
1. 先看索引or列名是否有意义
2. 先解决结构问题（三要求）
3. 再解决内容问题

63 对索引和列名重命名（**使用字典,老键新值**）

In [19]:
df_63 = pd.read_csv(r'D:\data analysis\LinLiLi\04数据评估与清洗篇\example1.csv', index_col=0)
df_63

客户_姓名 客户 性别  age                    邮箱
1     李十     女   58  yVaQXekW@example.com
2_    冯三     男   27  LctXbrEM@example.com
3     吴十     男   32  UyTYBzUZ@example.com
4     王十     女   58  KCixgciF@example.com
_5    钱十     女   27  nqRBXOtA@example.com
6*    赵八     女   51  GyKwILAL@example.com

In [20]:
# rename()传入字典，不改变原始dataframe
df_63.rename(index={"2_":"2", "_5":"5", "6*":"6"},
             columns={"客户_姓名":"客户姓名", "客户 性别":"客户性别", "age":"客户年龄", "邮箱":"客户邮箱"})


客户姓名 客户性别  客户年龄                  客户邮箱
1   李十    女    58  yVaQXekW@example.com
2   冯三    男    27  LctXbrEM@example.com
3   吴十    男    32  UyTYBzUZ@example.com
4   王十    女    58  KCixgciF@example.com
5   钱十    女    27  nqRBXOtA@example.com
6   赵八    女    51  GyKwILAL@example.com

- 可以传入一个针对Series的函数或方法,把原本索引or列名作为参数，然后返回新的名字。   
`df_63.rename(index=某函数/方法,
             columns=某函数/方法)`  
`df_63.rename(index=str.upper)` 

- `.set_index(),sort_index()`

64 ：清理结构性问题
- 对列进行拆分

In [21]:
df_64 = pd.read_csv(r'D:\data analysis\LinLiLi\04数据评估与清洗篇\example5.csv')
df_64

城市              人口密度
0  城市A  28.53万人/0.87平方公里
1  城市B  37.83万人/2.19平方公里
2  城市C   15.3万人/1.57平方公里

In [27]:
df_64[["人口", "面积"]] = df_64["人口密度"].str.split("/", expand=True)
df_64

城市              人口密度       人口        面积
0  城市A  28.53万人/0.87平方公里  28.53万人  0.87平方公里
1  城市B  37.83万人/2.19平方公里  37.83万人  2.19平方公里
2  城市C   15.3万人/1.57平方公里   15.3万人  1.57平方公里

In [29]:
df_64 = df_64.drop("人口密度", axis=1)
df_64

城市       人口        面积
0  城市A  28.53万人  0.87平方公里
1  城市B  37.83万人  2.19平方公里
2  城市C   15.3万人  1.57平方公里

- 合并列

In [53]:
df_642= pd.DataFrame({'姓': ['张', '李', '王'], '名': ['三', '四', '五'], '年龄': [30, 25, 35]})
df_642

姓  名  年龄
0  张  三  30
1  李  四  25
2  王  五  35

In [54]:
df_642["姓"].str.cat(df_642["名"], sep="-")

0    张-三
1    李-四
2    王-五
Name: 姓, dtype: object

In [55]:
df_642['姓名'] = df_642["姓"].str.cat(df_642["名"])

In [46]:
df_642 = df_642.drop(['姓','名'],axis=1)
df_642

年龄  姓名
0  30  张三
1  25  李四
2  35  王五

- 把宽数据转换成长数据

In [57]:
df_643 = pd.read_csv(r'D:\data analysis\LinLiLi\04数据评估与清洗篇\example6.csv')
df_643

国家代码    年份  男性年龄组（0-4岁）  男性年龄组（5-14岁）  男性年龄组（15-24岁）  女性年龄组（0-4岁）  \
0   CN  2019          100           200            500           80   
1   US  2019           50           150            300           40   
2   JP  2019           30           120            250           20   
3   IN  2019           80           180            400           60   

   女性年龄组（5-14岁）  女性年龄组（15-24岁）  
0           150            400  
1           100            200  
2            80            150  
3           120            300

In [58]:
df_643 = pd.melt(df_643,
                 id_vars=['国家代码', '年份'],    # 保持不变的列
                 var_name='年龄组',              # 新列的名称
                 value_name='肺结核病例数'       # 新列对应的值
)
df_643

国家代码    年份            年龄组  肺结核病例数
0    CN  2019    男性年龄组（0-4岁）     100
1    US  2019    男性年龄组（0-4岁）      50
2    JP  2019    男性年龄组（0-4岁）      30
3    IN  2019    男性年龄组（0-4岁）      80
4    CN  2019   男性年龄组（5-14岁）     200
5    US  2019   男性年龄组（5-14岁）     150
6    JP  2019   男性年龄组（5-14岁）     120
7    IN  2019   男性年龄组（5-14岁）     180
8    CN  2019  男性年龄组（15-24岁）     500
9    US  2019  男性年龄组（15-24岁）     300
10   JP  2019  男性年龄组（15-24岁）     250
11   IN  2019  男性年龄组（15-24岁）     400
12   CN  2019    女性年龄组（0-4岁）      80
13   US  2019    女性年龄组（0-4岁）      40
14   JP  2019    女性年龄组（0-4岁）      20
15   IN  2019    女性年龄组（0-4岁）      60
16   CN  2019   女性年龄组（5-14岁）     150
17   US  2019   女性年龄组（5-14岁）     100
18   JP  2019   女性年龄组（5-14岁）      80
19   IN  2019   女性年龄组（5-14岁）     120
20   CN  2019  女性年龄组（15-24岁）     400
21   US  2019  女性年龄组（15-24岁）     200
22   JP  2019  女性年龄组（15-24岁）     150
23   IN  2019  女性年龄组（15-24岁）     300

- 行拆分

In [65]:
df_644 = pd.read_csv(r'04数据评估与清洗篇/example7.csv')
df_644

学生姓名  学号                      课程列表
0   张三   1              ['数学', '物理']
1   李四   2        ['英语', '化学', '历史']
2   王五   3  ['语文', '数学', '英语', '政治']
3   赵六   4              ['物理', '生物']

虽然上面的课程列表变量看起来像是列表，但实际上是字符串  
所以我们需要先将字符串形式的列表转换为实际的列表对象，这样explode才能对列表进行拆分

In [66]:
# 使用Python内置的`eval`函数，它可以把字符串转换成表达式，所以可以把表示列表的字符串转换成列表本身
df_644['课程列表'] = df_644['课程列表'].apply(lambda x: eval(x))
df_644

学生姓名  学号              课程列表
0   张三   1          [数学, 物理]
1   李四   2      [英语, 化学, 历史]
2   王五   3  [语文, 数学, 英语, 政治]
3   赵六   4          [物理, 生物]

In [68]:
df_644.explode('课程列表')

学生姓名  学号 课程列表
0   张三   1   数学
0   张三   1   物理
1   李四   2   英语
1   李四   2   化学
1   李四   2   历史
2   王五   3   语文
2   王五   3   数学
2   王五   3   英语
2   王五   3   政治
3   赵六   4   物理
3   赵六   4   生物

- 删除行or列。.drop()

65 ：清理内容性问题

- 缺失数据
1. 人工寻找，然后赋值填补。loc和iloc
2. 自动填补。.fillna(value=)。series和dataframe都有此方法

In [70]:
import numpy as np
df_65 = pd.DataFrame({'A': [1, 2, np.nan, 4],
                   'B': [5, np.nan, 7, np.nan],
                   'C': [8, 9, 10, 11]})
df_65

A    B   C
0  1.0  5.0   8
1  2.0  NaN   9
2  NaN  7.0  10
3  4.0  NaN  11

In [71]:
df_65.fillna({'A': 0, 'B': 10, 'C': 20})

A     B   C
0  1.0   5.0   8
1  2.0  10.0   9
2  0.0   7.0  10
3  4.0  10.0  11

- 删除存在缺失值的行/列 .drop()

65  
`pd.to_csv()` 默认会把索引保存，可以index=False忽略索引；  
`pd.read_csv()`读取时，不知道那是索引，会生成列名为unnamed:0的第一列  
故用rename()，重命名；然后set_index()

# 66、67项目：评估和清理英国电商公司销售数据

目标：  
此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。   
本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。  
简介：  
原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [77]:
df_66 = pd.read_csv(r'D:\data analysis\LinLiLi\04数据评估与清洗篇\e_commerce.csv')
df_66.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据
- 结构，整齐度
- 内容，干净度。  
数据的结构性问题指:不符**每列是一个变量，每行是一个观察值，每个单元格是一个**这三个标准，  
数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

- 评估数据整齐度

In [78]:
df_66.sample(10)

InvoiceNo StockCode                        Description  Quantity  \
382724    569943     22966      GINGERBREAD MAN COOKIE CUTTER         1   
186351    552857     22930  BAKING MOULD HEART MILK CHOCOLATE         1   
299866    563175     23054     RECYCLED ACAPULCO MAT LAVENDER         6   
378545    569653     21981       PACK OF 12 WOODLAND TISSUES         24   
157307    550201     22629                SPACEBOY LUNCH BOX         12   
533345    581023     22327  ROUND SNACK BOXES SET OF 4 SKULLS         2   
44213     540159     84817       DANISH ROSE DECORATIVE PLATE        48   
319559    564857     22179            SET 10 NIGHT OWL LIGHTS         1   
148734    549251     21087            SET/6 POSIES PAPER CUPS         1   
462463    575975     22617         BAKING SET SPACEBOY DESIGN         3   

             InvoiceDate  UnitPrice  CustomerID         Country  
382724   10/6/2011 18:08       1.25     14592.0  United Kingdom  
186351   5/11/2011 15:34       4.96         NaN  United Kingdom  
299866   8/12/2011 14:02       8.25     14855.0  United Kingdom  
378545   10/5/2011 12:52       0.39     12451.0     Switzerland  
157307   4/15/2011 10:25       1.95     12769.0          Brazil  
533345   12/7/2011 10:35       6.63         NaN  United Kingdom  
44213     1/5/2011 11:47       0.42     16112.0  United Kingdom  
319559    8/31/2011 9:23      13.29         NaN  United Kingdom  
148734    4/7/2011 12:16       0.19     14449.0  United Kingdom  
462463  11/13/2011 13:18       4.95     13536.0  United Kingdom

从抽样的10行数据数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

- 评估数据干净度

In [79]:
df_66.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，而`Description`、`CustomerID`变量存在缺失值。  
此外，`InvoiceDate`的数据类型应为日期，`CustomerID`的数据类型应为字符串，应当进行数据格式转换。

$$ 1. 评估缺失数据$$   
- 在了解`Description`存在缺失值后，根据条件提取出缺失观察值。

In [86]:
df_66[df_66['Description'].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易数据缺失`Description`变量值。

从输出结果来看，这些缺失`Description`的交易数据，`UnitPrice`都为0。为了验证猜想，我们增加筛选条件，看是否存在`Description`变量缺失且`UnitPrice`不为0的数据。

In [91]:
df_66[(df_66['Description'].isnull()) & (df_66['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来结果数量为0条，说明缺失`Description`值的数据，同时也不具备有效的`UnitPrice`值。  
`Description`表示产品名称，`UnitPrice`表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

- `CustomerID`变量同样存在缺失值，因此也根据条件提取出缺失观察值。

In [93]:
df_66[df_66['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

`CustomerID`表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些`CustomerID`缺失的交易数据仍然有效，因此保留此变量为空的观察值。


$$ 2. 评估重复数据$$
根据数据变量的含义来看，虽然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一标识符，但一次交易可能包含多件商品，因此`InvoiceNo`可以存在重复，不同交易可以包含同件商品，因此`StockCode`可以存在重复，顾客可以进行多次交易或下单多个商品，因此`CustomerID`也可以存在重复。

那么针对此数据集，我们无需评估重复数据。

$$ 3. 评估不一致数据$$
不一致数据可能存在于`Country`变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [95]:
df_66['Country'].value_counts()

United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68
European

从`Country`变量值来看，`"USA"`、`"United States"`均在表示美国，`"United Kingdom"`、`"UK"`、`"U.K."`均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

$$4. 评估无效或错误数据$$
可以通过DataFrame的`describe`方法，对数值统计信息进行快速了解。

In [96]:
df_66.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，`Quantity`和`UnitPrice`存在负数，会对后续数值分析产生影响。
- 因此，我们先筛选出`Quantity`数值为负数的观察值，进一步评估其含义。

In [98]:
df_66[df_66.Quantity <0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，`Quantity`变量为负数的观察值，`InvoiceNo`似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

In [101]:
df_66[(df_66.Quantity <0) & (df_66.InvoiceNo.str[0]!= "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

但根据以上输出结果，这些筛选出的观察值的`UnitPrice`观察值均为0，因此增加`UnitPrice`的条件进行验证。

In [103]:
df_66[(df_66.Quantity <0) & (df_66.InvoiceNo.str[0]!= "C") & (df_66.UnitPrice != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当`Quantity`变量为负数的时候，观察值满足以下条件之一：
>`InvoiceNo`以C开头，表示订单被取消
>>`UnitPrice`为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将`Quantity`为负数的观察值删除。

- 接下来筛选出`UnitPrice`为负数的观察值。

In [104]:
df_66[df_66.UnitPrice < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负数的观察值都是坏账调账，不属于实际商品交易数据，因此也在数据清理步骤中也将其删除。

## 清理数据
根据前面评估部分得到的结论，我们需要进行的数据清理包括：
- 把`InvoiceDate`变量的数据类型转换为为日期时间
- 把`CustomerID`变量的数据类型转换为字符串
- 把`Description`变量缺失的观察值删除
- 把`Country`变量值`"USA"`替换为`"United States"`
- 把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`
- 把`Quantity`变量值为负数的观察值删除
- 把`UnitPrice`变量值为负数的观察值删除

为了区分开经过清理的数据和原始的数据，我们创建新的变量`cleaned_data`，让它为复制出的副本。我们之后的清理步骤都将被运用在`cleaned_data`上。


In [106]:
cleaned_data = df_66.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

- 把`InvoiceDate`变量的数据类型转换为为日期时间类型：


In [108]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'])
cleaned_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

In [110]:
cleaned_data["CustomerID"] = df_66["CustomerID"].astype(str) # pandas里字符串类型叫object
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

In [115]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0,-2)
cleaned_data["CustomerID"]

0         178
1         178
2         178
3         178
4         178
         ... 
541904    126
541905    126
541906    126
541907    126
541908    126
Name: CustomerID, Length: 541909, dtype: object

In [116]:
cleaned_data.dropna(subset=["Description"], inplace=True)
cleaned_data["Description"].isnull().sum()

0

In [117]:
cleaned_data["Country"] = df_66["Country"].replace({"USA": "United States"})
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

In [118]:
cleaned_data["Country"] = df_66["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


In [121]:
cleaned_data = df_66[df_66["Quantity"] >= 0]
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

In [122]:
cleaned_data = df_66[df_66["UnitPrice"] >= 0]
len(cleaned_data[cleaned_data["UnitPrice"] < 0])


0

## 保存数据

In [123]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [125]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False) #索引有意义时才需要保存
pd.read_csv("e_commerce_cleaned.csv").head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

# 74 75 76统计学（描述和推断）

- 分类数据（定类和定序）
- **数值数据**（连续和离散）
  > 集中趋势：平均数，中位数，众数`df['体温']).mode()`，
  >>离散趋势：极差`np.array(df['体温']).ptp()`，方差`df['体温']).var()`标准差，   
                四分位距，第三`df['体温']).quantile(0.75)`-第一`df['体温']).quantile(0.25)`，中间50%数据，距离越大越离散。 
  >>>分布形状：分布对称程度；峰值高低程度。
  
  > 正偏态，右偏，长尾巴在右边, 大部分数值低于平均值。
  >>负偏态，左偏，长尾巴在左边，大部分数值高于于平均值

# 数据可视化

77 一个变量

78 多个变量